ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,536 @@
+            plugin.onDefaultPhase(plugin.fallbackInterceptions, ApplicationCallPipeline.Fallback, block)
+        }
+    }
+
+    /**
+     * Callable object that defines how receiving data from HTTP call should be modified by the current [KtorPlugin].
+     * */
+    public override val onCallReceive: OnReceive = object : OnReceive {
+        private val plugin = this@KtorPlugin
+
+        override fun invoke(block: suspend ReceiveExecution.(ApplicationCall) -> Unit) {
+            plugin.onDefaultPhase(plugin.onReceiveInterceptions, ApplicationReceivePipeline.Transform, block)
+        }
+
+        override fun beforeTransform(block: suspend ReceiveExecution.(ApplicationCall) -> Unit) {
+            plugin.onDefaultPhase(plugin.beforeReceiveInterceptions, ApplicationReceivePipeline.Before, block)
+        }
+    }
+
+    /**
+     * Callable object that defines how sending data to a client within HTTP call should be modified by the current [KtorPlugin].
+     * */
+    public override val onCallRespond: OnRespond = object : OnRespond {
+        private val plugin = this@KtorPlugin
+
+        override fun invoke(block: suspend ResponseExecution.(ApplicationCall) -> Unit) {
+            plugin.onDefaultPhase(plugin.onResponseInterceptions, ApplicationSendPipeline.Transform, block)
+        }
+
+        override fun beforeTransform(block: suspend ResponseExecution.(ApplicationCall) -> Unit) {
+            plugin.onDefaultPhase(plugin.beforeResponseInterceptions, ApplicationSendPipeline.Before, block)
+        }
+
+        override fun afterTransform(block: suspend ResponseExecution.(ApplicationCall) -> Unit) {
+            plugin.onDefaultPhase(plugin.afterResponseInterceptions, ApplicationSendPipeline.After, block)
+        }
+    }
+
+    public abstract class RelativePluginContext(
+        private val currentPlugin: InterceptionsHolder,
+        private val otherPlugin: InterceptionsHolder
+    ) : PluginContext {
+        protected fun <T : Any> sortedPhases(
+            interceptions: List<Interception<T>>,
+            pipeline: Pipeline<*, ApplicationCall>
+        ): List<PipelinePhase> =
+            interceptions
+                .map { it.phase }
+                .sortedBy {
+                    if (!pipeline.items.contains(it)) {
+                        throw PluginNotInstalledException(otherPlugin.name)
+                    }
+
+                    pipeline.items.indexOf(it)
+                }
+
+        public abstract fun selectPhase(phases: List<PipelinePhase>): PipelinePhase?
+
+        public abstract fun insertPhase(
+            pipeline: Pipeline<*, ApplicationCall>,
+            relativePhase: PipelinePhase,
+            newPhase: PipelinePhase
+        )
+
+        private fun <T : Any> insertToPhaseRelatively(
+            currentInterceptions: MutableList<Interception<T>>,
+            otherInterceptions: MutableList<Interception<T>>,
+            block: suspend Execution<T>.(ApplicationCall) -> Unit
```

Comment:
It looks like we just renamed `PipelineContext` to `Execution`. Actually `PipelineContext` for me has more sense than `Execution`. Could you explain what `Execution` is and how are we going to explain this in docs?